.github/workflows/build-and-test.yml

```diff
@@ -13,31 +13,29 @@ jobs:
 
     steps:
       - uses: actions/checkout@v2
+      - name: Link shellcheck into expected location
+        run: sudo ln -s $(which shellcheck) /usr/local/bin/shellcheck
+      - name: Set up Python
+        uses: actions/setup-python@v1
+        with:
+          python-version: 3.9
+      - name: Install Black
+        run: pip install black
```

Comment:
I'm not sure we should block the Java build on installing and running `shellcheck`, `Python`, and `black` first. My preference is to keep the linting step and not make `shellcheck` / `black` a mandatory part of the Java build.

Reply:
I don't see this as blocking the java build; my thinking is that the top level build is now building all gradle subprojects. This is similar to the `unit-tests-and-sonarqube` step, which runs the `test` task for all subprojects.

Running all builds in a single step means that we don't need to add a new step for each gradle subproject. It does have the downside that a lint error in the python code will block the java unit tests from running. The alternative is that the workflow code will need to be very specific about java build vs everything else, and java test vs everything else.

Reply:
Performance-wise this change has a negligible effect on the `build` job; the job completes in less than two minutes.

Reply:
> I think it does effectively block the Java build: the action steps run sequentially and require the installation of shellcheck in a specific place and black before the build can proceed, and then the build uses those tools to run the linting. Ideally we would be able to do this in parallel to the main build, which is similar to what the linters job did beforehand. 

This is true, but what would be the gain of this change? Right now the the performance difference is negligible (less than 1 minute). An error in one area (like shellcheck) might block detection of errors in another area (like black) but once a user knows there's a compilation error, they can iterate locally. The downside is that the GitHub workflow would be more complex, which means harder to maintain. Keeping the workflow less complex seems more important to me than making these steps parallel. It's possible gradle supports this kind of parallelism, so there might be a way to fix this problem without adding complexity.

> requiring shellcheck to exist at a particular location for the build to run seems unusual to me.

This is a separate issue. The location is configurable but I don't know if there's a way to find it on the path and then pass that location to the plugin. The plugin can run `shellcheck` inside docker, which avoids this issue, but there's an overhead to launching docker that I'd like to avoid if possible. Looking at the [docs](https://github.com/felipefzdz/gradle-shellcheck-plugin#the-shellcheck-plugin), I think using `SHELLCHECK_PATH` would avoid the need to link shellcheck into place, so I can look into that too.

> I agree that running all builds in a single step is a convenience as a user, but I don't necessarily see that as an advantage in a GitHub Action where we may need to build specific parts of the project for specific cases (client vs server code, different types of scripts, etc)

That's true, it's not necessarily an advantage, although so far we don't have a need to build specific parts in our action.

> Additionally, as a user if I am making Java code change I shouldn't need to scan all my Python and shell scripts if those haven't changed

Gradle caches the results of a task run and shouldn't re-run the tasks unless their sources have changed. So this change shouldn't affect local development in this way.